### DISCUSSION TITLE
Prisma client extension utilizing $allModels to transform the where input.

### DISCUSSION THREAD

**Discussion by jbeckton** [prisma/prisma#23932] [Q&A]
### Question

I'm struggling to put together a client extension that can transform the where input for the query.   

I have the following working just fine...

```` typescript

export const extendPrismaClientFor = (currentUser: CurrentUserContext) => {
  return Prisma.defineExtension({
    name: 'extendedClient',
    query: {
      user: {
        async findMany({ model, operation, args, query }) {
          let enhancedWhereInput = {};
          if (currentUser?.abilities) {
            enhancedWhereInput = accessibleBy(currentUser.abilities).User;
          }
          const newArgs: Prisma.UserFindManyArgs = { ...args, where: { ...args.where, AND: [enhancedWhereInput] } };
          return query(newArgs);
        },
      },
    },
  });
};

````

When I try and do this using `$allModels` on the `findMany` operation the arg types are difficult to figure out. Type checking won't allow me to transform the `where` either. 

How do I set the proper type for the args when the current model is unknown?

Thanks.

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @jbeckton 

I've shared this internally. We will get back to you with updates.

**Reply by ludralph** [reply]
Hi @jbeckton 

When using $allModels, you need to use generic types to handle the unknown model. Here's an example of how you can modify your extension to work with $allModels:
```ts
import { Prisma } from '@prisma/client';

export const extendPrismaClientFor = (currentUser: CurrentUserContext) => {
  return Prisma.defineExtension({
    name: 'extendedClient',
    query: {
      $allModels: {
        async findMany<T, A>(
          this: T,
          args: Prisma.Exact<A, Prisma.Args<T, 'findMany'>>
        ): Promise<Prisma.Result<T, A, 'findMany'>> {
          const context = Prisma.getExtensionContext(this);
          let enhancedWhereInput = {};
          if (currentUser?.abilities) {
            enhancedWhereInput = accessibleBy(currentUser.abilities)[context.$name];
          }
          const newArgs = {
            ...args,
            where: {
              ...args.where,
              AND: [enhancedWhereInput]
            }
          };
          return (context as any).findMany(newArgs);
        },
      },
    },
  });
};
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;
